src/Middleware/Diagnostics/src/StatusCodePage/StatusCodePagesExtensions.cs

```diff
@@ -210,7 +210,32 @@ public static IApplicationBuilder UseStatusCodePages(this IApplicationBuilder ap
                     context.HttpContext.Request.Path = originalPath;
                     context.HttpContext.Features.Set<IStatusCodeReExecuteFeature?>(null);
                 }
-            });
+            };
+
+            const string globalRouteBuilderKey = "__GlobalEndpointRouteBuilder";
+            // Check if UseRouting() has been called so we know if it's safe to call UseRouting()
+            // otherwise we might call UseRouting() when AddRouting() hasn't been called which would fail
+            if (app.Properties.TryGetValue("__EndpointRouteBuilder", out _) || app.Properties.TryGetValue(globalRouteBuilderKey, out _))
+            {
+                return app.Use(next =>
+                {
+                    app.Properties.TryGetValue(globalRouteBuilderKey, out var routeBuilder);
+                    // start a new middleware pipeline
+                    var builder = app.New();
+                    if (routeBuilder is not null)
+                    {
+                        // use the old routing pipeline if it exists so we preserve all the routes and matching logic
+                        builder.Properties[globalRouteBuilderKey] = routeBuilder;
+                    }
+                    builder.UseStatusCodePages(handler);
+                    builder.UseRouting();
+                    // apply the next middleware
+                    builder.Run(next);
+                    return builder.Build();
```

Comment:
This doesn't make sense to me. We only want to run the branch if we end up re-writing the path. We don't want to do it all the time. Otherwise, there's no point to branching, we could just run it inline and it would be no different. 

We need to add an internal property to the options for the re-execute path.